### ISSUE TITLE
ddl.dic: proposal to add a bit more semantics to the 'Key' _type.purpose

### ISSUE THREAD

**Issue by vaitkus** [COMCIFS/cif_core#108]
The ``ddl.dic`` dictionary defines the **Key** type purpose as:
```
Used to type an item with a value that is unique within
the looped list of these items, and may be used as a
reference "key" to identify a specific packet of items
within the category.
``` 

The definition seems to state that the data item is a candidate key within its category. However, the same functionality seems to also be implemented by the  ``_category.key_id`` and  ``_category_key.name`` data items.

I propose to slightly expand the semantics of the 'Key' purpose type by clearly stating that data items of this purpose act strictly as artificial primary keys. That is, they do not carry any additional semantics beyond being unique identifiers.

This ability to differentiate between natural and artificial natural keys would prove useful during the validation process. It is acceptable for artificial keys to be omitted under certain condition whereas the absence of natural keys data items tend to be a signal of incomplete data. For example, the `_diffrn_radiation_wavelength.id` and `_atom_type_scat.symbol` data items both act as primary keys of their categories. Absence of the `_atom_type_scat.symbol` data item impedes one from correctly interpreting atom scattering information, whereas absence of the `_diffrn_radiation_wavelength.id` might be tolerable in data files with a single wavelength or those not containing additional references to the data item (i.e. the `_diffrn_refln.wavelength_id` data item).

An undesired side effect of this change is the loss of means to enforce uniqueness constraint on a single non-key data item. However, I would also argue that this constraint should be specified as a separate property of the data item rather than be coupled in with its purpose.

**Comment by jamesrhester** [comment]
This is an interesting idea that I like. mmCIF routinely assumes the value of artificial keys when there is a single row in a category. I will run it by the DDLm group to make sure they are OK with it.

**Comment by jamesrhester** [comment]
One member of the DDLm group objected to the change on general principles, but was unable to come up with any specific scenario, so this change is accepted. The updated text presented to them was:

```
Key
;                  Used to type an item with a value that is unique within
                   the looped list of these items, and does not contain encoded
                   information. 
;
```

This can now be incorporated into ddl.dic and the relevant datanames updated.

**Comment by vaitkus** [comment]
In order to modify the definition without making the dictionaries contradict themselves some additional changes need to be made.

Currently, the DDL dictionary contains 8 data item definitions with the 'Key' purpose. Most of them do not fit the new definition and need to be changed:

| Data name | Proposed new purpose  |
| ------------- |-------------|
| _alias.definition_id  | Encode | 
| _description_example.case | Describe |  
| _dictionary_audit.version | Encode | 
| _dictionary_valid.application | Encode |
| _dictionary_xref.code | Encode |
| _enumeration_default.index | Encode |
| _enumeration_set.state | Encode |
| _import_details.order |  Encode |

All usages of the 'Key' purpose in the CIF_CORE dictionary seem to fit the new definition.

If the proposed changes seem reasonable I can incorporate them into the dictionary and issue a PR.

**Comment by jamesrhester** [comment]
I agree with all of the changes suggested by @vaitkus .

**Comment by vaitkus** [comment]
Issue has been solved in PR #125.

**Comment by jamesrhester** [comment]
Thank you for your PR.
